### PR TITLE
Mx 15792 esdt prefix updates and fixes

### PR DESCRIPTION
**Pull request by axenteoctavian**
## Reasoning behind the pull request
- ESDT prefix fixes for some token types
- 
- 
  
## Proposed changes
- Cross-chain token checker in esdtDataStorage to bypass the change of token type
- Refactored tests, added cross-shard transactions, added multiple checks in user account and system account 
- 

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/disabled/crossChainTokenChecker.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package disabled
+
+type crossChainTokenChecker struct {
+}
+
+func NewCrossChainTokenChecker() *crossChainTokenChecker {
```
Comment (mariusmihaic) [inline factory/disabled/crossChainTokenChecker.go:6]: Missing comm

File: factory/disabled/crossChainTokenChecker.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	return &crossChainTokenChecker{}
+}
+
+// IsCrossChainOperation does nothing
```
Comment (mariusmihaic) [inline factory/disabled/crossChainTokenChecker.go:10]:
```suggestion
// IsCrossChainOperation returns false
```

File: factory/disabled/crossChainTokenChecker.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+func (ctc *crossChainTokenChecker) IsCrossChainOperation(tokenID []byte) bool {
+	return false
+}
+
+// IsCrossChainOperationAllowed does nothing
```
Comment (mariusmihaic) [inline factory/disabled/crossChainTokenChecker.go:15]:
```suggestion
// IsCrossChainOperationAllowed returns false nothing
```

File: factory/disabled/crossChainTokenChecker.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+type crossChainTokenChecker struct {
+}
+
+func NewCrossChainTokenChecker() *crossChainTokenChecker {
+	return &crossChainTokenChecker{}
+}
+
+// IsCrossChainOperation does nothing
+func (ctc *crossChainTokenChecker) IsCrossChainOperation(tokenID []byte) bool {
```
Comment (mariusmihaic) [inline factory/disabled/crossChainTokenChecker.go:11]:
```suggestion
func (ctc *crossChainTokenChecker) IsCrossChainOperation(_ []byte) bool {
```

File: factory/disabled/crossChainTokenChecker.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	return false
+}
+
+// IsCrossChainOperationAllowed does nothing
+func (ctc *crossChainTokenChecker) IsCrossChainOperationAllowed(address []byte, tokenID []byte) bool {
```
Comment (mariusmihaic) [inline factory/disabled/crossChainTokenChecker.go:16]:
```suggestion
func (ctc *crossChainTokenChecker) IsCrossChainOperationAllowed(_ []byte, _ []byte) bool {
```

File: factory/processing/processComponents.go [hunk 6]
```diff
@@ -585,11 +585,12 @@ func (pcf *processComponentsFactory) Create() (*processComponents, error) {
...
-		GlobalSettingsHandler: disabled.NewDisabledGlobalSettingHandler(),
-		Marshalizer:           pcf.coreData.InternalMarshalizer(),
-		ShardCoordinator:      pcf.bootstrapComponents.ShardCoordinator(),
-		EnableEpochsHandler:   pcf.coreData.EnableEpochsHandler(),
+		Accounts:                      pcf.state.AccountsAdapterAPI(),
+		GlobalSettingsHandler:         disabled.NewDisabledGlobalSettingHandler(),
+		Marshalizer:                   pcf.coreData.InternalMarshalizer(),
+		ShardCoordinator:              pcf.bootstrapComponents.ShardCoordinator(),
+		EnableEpochsHandler:           pcf.coreData.EnableEpochsHandler(),
+		CrossChainTokenCheckerHandler: disabled.NewCrossChainTokenChecker(),
```
Comment (mariusmihaic) [inline factory/processing/processComponents.go:593]: Why is this disabled?

File: integrationTests/chainSimulator/bridge/executeAndBurn_test.go [hunk 7]
```diff
@@ -259,40 +244,91 @@ func getBridgedValue(bridgeInTokens []chainSim.ArgsDepositToken, token string) (
...
+	nodeHandler process.NodeHandler,
+	token chainSim.ArgsDepositToken,
+) {
+	accountKeys, _, err := nodeHandler.GetFacadeHandler().GetKeyValuePairs(esdtSystemAccount, dataApi.AccountQueryOptions{})
+	require.Nil(t, err)
+	require.NotNil(t, accountKeys)
+	require.Empty(t, accountKeys[getTokenKey(token)])
+}
+
+func metaDataOnUserAccount(esdtType core.ESDTType) bool {
```
Comment (mariusmihaic) [inline integrationTests/chainSimulator/bridge/executeAndBurn_test.go:326]: This func just checks esdt type, could be renamed

File: integrationTests/chainSimulator/bridge/executeAndBurn_test.go [hunk 8]
```diff
@@ -72,19 +72,11 @@ func TestChainSimulator_ExecuteWithMintMultipleEsdtsAndBurnNftWithDeposit(t *tes
...
 	}
 
-	nft := sovChainPrefix + "-SOVNFT-123456"
-	nftNonce := uint64(3)
 	nftV2 := sovChainPrefix + "-NFTV2-a1b2c3"
 	nftV2Nonce := uint64(10)
 	token := sovChainPrefix + "-TKN-1q2w3e"
 
 	bridgedInTokens := make([]chainSim.ArgsDepositToken, 0)
-	bridgedInTokens = append(bridgedInTokens, chainSim.ArgsDepositToken{
```
Comment (mariusmihaic) [inline integrationTests/chainSimulator/bridge/executeAndBurn_test.go:82]: Why were these removed from these tests?

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go [hunk 9]
```diff
@@ -38,7 +39,7 @@ func TestChainSimulator_ExecuteMintBurnBridgeOpForESDTTokensWithPrefixAndTransfe
 		BypassTxSignatureCheck:   true,
 		TempDir:                  t.TempDir(),
 		PathToInitialConfig:      defaultPathToInitialConfig,
-		NumOfShards:              1,
+		NumOfShards:              3,
```
Comment (mariusmihaic) [inline integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:42]: :+1:

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go [hunk 10]
```diff
@@ -79,57 +81,144 @@ func TestChainSimulator_ExecuteMintBurnBridgeOpForESDTTokensWithPrefixAndTransfe
...
-
-	// We will deposit a prefixed token from a sovereign chain to the main chain,
-	// expecting these tokens to be minted by the whitelisted ESDT safe sc and transferred to our address.
+	bridgedInTokens = append(bridgedInTokens, chainSim.ArgsDepositToken{
+		Identifier: argsEsdtSafe.ChainPrefix + "-NFTV2-1q2w3e",
+		Nonce:      uint64(3),
+		Amount:     big.NewInt(1),
+		Type:       core.NonFungibleV2,
+	})
+	//bridgedInTokens = append(bridgedInTokens, chainSim.ArgsDepositToken{
```
Comment (mariusmihaic) [inline integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:98]: This is not ok.
We should not merge commented untested code !

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go [hunk 11]
```diff
@@ -79,57 +81,144 @@ func TestChainSimulator_ExecuteMintBurnBridgeOpForESDTTokensWithPrefixAndTransfe
...
+		// expecting these tokens to be burned by the whitelisted ESDT safe sc
+		txResult = Deposit(t, cs, receiver.Bytes, &receiverNonce, bridgeData.ESDTSafeAddress, []chainSim.ArgsDepositToken{bridgedOutToken}, receiver.Bytes)
+		chainSim.RequireSuccessfulTransaction(t, txResult)
+
+		remainingTokenAmount := big.NewInt(0).Sub(receiverToken.Amount, amountToDeposit)
+		checkTokenDataInAccount(t, cs, token, receiver.Bech32, remainingTokenAmount)
+		checkTokenDataInAccount(t, cs, token, esdtSafeEncoded, big.NewInt(0))
+		receiverToken.Amount = remainingTokenAmount
+
+		if remainingTokenAmount.Cmp(big.NewInt(0)) > 0 {
```
Comment (mariusmihaic) [inline integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:182]: Why make this "same shard" transfer based on remaining amount?

You seem to always send tokens above from `receiver`, then in "same shard" you also send from `receiver`, so you should always have tokens in `receiver`

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go [hunk 12]
```diff
@@ -79,57 +81,144 @@ func TestChainSimulator_ExecuteMintBurnBridgeOpForESDTTokensWithPrefixAndTransfe
...
+	token chainSim.ArgsDepositToken,
+	sender, receiver dtos.WalletAddress,
+	senderNonce *uint64,
+	amountToSend *big.Int,
+) chainSim.ArgsDepositToken {
+	if token.Type == core.Fungible {
+		chainSim.TransferESDT(t, cs, sender.Bytes, receiver.Bytes, senderNonce, token.Identifier, amountToSend)
+	} else {
+		chainSim.TransferESDTNFT(t, cs, sender.Bytes, receiver.Bytes, senderNonce, token.Identifier, token.Nonce, amountToSend)
+		err := cs.GenerateBlocks(3)
```
Comment (mariusmihaic) [inline integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:203]: Why generate extra blocks here? 
Because cross shard txs? And why it works for `TransferESDT` without generated blocks?

**Review by mariusmihaic — CHANGES_REQUESTED** [review]
We should not merge commented untested code

**Review by BeniaminDrasovean — COMMENTED** [review]
Just what Marius said

File: integrationTests/chainSimulator/bridge/executeAndBurn_test.go (hunk 8 above)
Comment (axenteoctavian) [reply integrationTests/chainSimulator/bridge/executeAndBurn_test.go:82]: `NonFungible` type is not accepted anymore

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go (hunk 11 above)
Comment (axenteoctavian) [reply integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:182]: - The `wallet` in shard 0 sends tokens to `receiver` in shard 2
- The `receiver` will deposit (burn) in the bridge contract in shard 1
- The `receiver` in shard 2 sends tokens to `receiver2` in same shard
This check is to skip the NFT tokens because after deposit quantity is 0

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go (hunk 12 above)
Comment (axenteoctavian) [reply integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:203]: For NFT transfer, sender == receiver in arguments and actual receiver is in input data in tx. So the chain simulator verifies the status on the same shard and is successful in 1 block. Then I need to wait 3 more blocks so the actual receiver will receive the tokens because is in different shard.

File: integrationTests/chainSimulator/bridge/executeAndBurn_test.go (hunk 7 above)
Comment (axenteoctavian) [reply integrationTests/chainSimulator/bridge/executeAndBurn_test.go:326]: It is used in the context of checking metadata. Checks if metadata is on user account based on esdt type. 
Any suggestions of name?

File: factory/processing/processComponents.go (hunk 6 above)
Comment (axenteoctavian) [reply factory/processing/processComponents.go:593]: The real component is not used here. It is used only in vm.

**Review by BeniaminDrasovean — DISMISSED** [review]

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go (hunk 11 above)
Comment (mariusmihaic) [reply integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:182]: Can we make the check based on type then?
E.g.: if the remaining amount is zero and the type is nft => correct behavior, otherwise fail, because this might hide possible bugs

File: factory/processing/processComponents.go (hunk 6 above)
Comment (mariusmihaic) [reply factory/processing/processComponents.go:593]: I still don't think this is correct.

**Review by mariusmihaic — COMMENTED** [review]
Just remaining open questions

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go (hunk 11 above)
Comment (axenteoctavian) [reply integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:182]: As we discusses in private, with the last commit this should be fixed :D

File: factory/processing/processComponents.go (hunk 6 above)
Comment (axenteoctavian) [reply factory/processing/processComponents.go]: created a real `crossChainTokenChecker` component

File: integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go [hunk 13]
```diff
@@ -79,57 +81,144 @@ func TestChainSimulator_ExecuteMintBurnBridgeOpForESDTTokensWithPrefixAndTransfe
 	paymentTokenAmount, _ := big.NewInt(0).SetString("1000000000000000000", 10)
 	chainSim.SetEsdtInWallet(t, cs, wallet, argsEsdtSafe.IssuePaymentToken, 0, esdt.ESDigitalToken{Value: paymentTokenAmount})
 
-	tokenToBridge := argsEsdtSafe.ChainPrefix + "-SOVT-5d8f56"
-	expectedMintValue := big.NewInt(123)
-
+	// TODO uncomment after rust framework and contract framework will be updated with all esdt types
```
Comment (axenteoctavian) [inline integrationTests/chainSimulator/bridge/bridgeAndTransfer_test.go:84]: The PR for this
https://github.com/multiversx/mx-chain-go/pull/6478

**Review by mariusmihaic — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by mariusmihaic — APPROVED** [review]
